### ISSUE TITLE
Flaky test: DockerSuite.TestEventsOOMDisableFalse

### ISSUE THREAD

**Issue by tiborvass** [docker/docker#17215]
Encountered on #17188

```
----------------------------------------------------------------------
FAIL: docker_cli_events_unix_test.go:57: DockerSuite.TestEventsOOMDisableFalse

docker_cli_events_unix_test.go:72:
    c.Assert(err, check.IsNil)
... value *errors.errorString = &errors.errorString{s:"wrong exit code for OOM container: expected 137, got 1 (output: \"Error response from daemon: Cannot start container 8e7c2324c27e8120caf40fbc97e78630b3d12436c0d3e509a12f871b84c08335: [8] System error: read parent: connection reset by peer\\n\")"} ("wrong exit code for OOM container: expected 137, got 1 (output: \"Error response from daemon: Cannot start container 8e7c2324c27e8120caf40fbc97e78630b3d12436c0d3e509a12f871b84c08335: [8] System error: read parent: connection reset by peer\\n\")")


----------------------------------------------------------------------
```

**Comment by clnperez** [comment]
FYI, I'm looking at these OOM tests since they do seem to be flaky (and are giving the gccgo run issues). So far I've just got a new test written and a few others modified, mostly exploring what all works and how. Is anyone else working on a PR for these? if not, #dibs. :)

**Comment by laboger** [comment]
I don't think the problem with this test is related to gccgo at all, it has to do with how the kernel on your test system is set up.

If I run the docker command from TestEventsOOMDisableFalse by hand on my test system I get this error:

docker run --name oomFalse -m 4MB ppc64le/busybox sh -c "x=a; while true; do x=$x$x$x$x; done"
WARNING: Your kernel does not support swap limit capabilities, memory limited without swap.

Which is documented here: 
http://docs.docker.com/engine/installation/ubuntulinux/
Under the heading Adjust memory and swap accounting

And sounds like if your kernel is not configured properly then the -m argument with 'docker run' is ignored.  That would explain the flaky behavior because it will behave differently depending on whether your kernel set up as expected.

Couldn't the test check for this message and provide a meaningful response?

**Comment by clnperez** [comment]
@tiborvass didn't open this specifically against gccgo as far as I can tell.

I am surprised that there's not a testRequires(memoryLimitSupport), given this uses the -m flag. But I know that the kernel for the PR systems was built with CONFIG_MEMCG_SWAP_ENABLED left as default (enabled), so I don't think that's the culprit for this case.

It's probably a good idea to add the testRequires, regardless.

**Comment by laboger** [comment]
Right, this was not originally written against gccgo, but was identified as a gccgo problem by Srini and is skipped when using gccgo.  To me the flaky behavior of this error that I've seen is consistent with ignoring the memory limit.

**Comment by mavenugo** [comment]
hitting the same issue in https://github.com/docker/docker/pull/19242 : https://jenkins.dockerproject.org/job/Docker-PRs-experimental/13670/console

**Comment by thaJeztah** [comment]
Are these resolved by https://github.com/docker/docker/pull/17913 and https://github.com/docker/docker/pull/20627?

**Comment by sainath14** [comment]
TestEventsOOMDisableFalse consistenly fails on my system with "Error: Timeout waiting for container to die on OOM" when CONFIG_MEMCG_SWAP_ENABLED is not set. As it takes around 7-8 minutes to reach OOM condition. But with swap enabled it would crash within couple of seconds. If this test case is intended to just test oom state for a container we might need to enforce testRequires to have swapLimit as condition instead of just memoryLimit.

**Comment by sean-jc** [comment]
The kernel config problem @laboger and @sainath14 are running into is due to a subtle dependency on --memory-swap defaulting to double the value of -m / --memory.  The OOM killer doesn't kick in until the container runs out of memory and swap.  When the memory swap cgroup is disabled the swap limit for the container will be the system swap limit, and will likely be many GB in size.  The tests timeout before they hit the system swap limit.

The hidden cgroup dependency was fixed by #20627 for TestEventsOOMDisableFalse, but there are a few other tests that need the swapMemorySupport requirement.  I'll create a PR to fix the relevant tests and to document the default behavior of --memory-swap for the docker run man page.

The original Jenkins failure appears to be something completely different; it might be variant of the GCCGO issue where the container could not even be started due to the memory limits being too restrictive.

**Comment by LK4D4** [comment]
Didn't see for a long time.
